sdl_android/src/main/java/com/smartdevicelink/api/screen/TextAndGraphicManager.java

```diff
@@ -0,0 +1,893 @@
+		pendingHMIFull = false;
+		textAlignment = CENTERED;
+		currentHMILevel = HMILevel.HMI_NONE;
+		currentScreenData = new Show();
+		addListeners();
+		getBlankArtwork();
+		transitionToState(READY);
+	}
+
+	@Override
+	public void dispose(){
+
+		textField1 = null;
+		textField1Type = null;
+		textField2 = null;
+		textField2Type = null;
+		textField3 = null;
+		textField3Type = null;
+		textField4 = null;
+		textField4Type = null;
+		mediaTrackTextField = null;
+		textAlignment = null;
+		primaryGraphic = null;
+		secondaryGraphic = null;
+		blankArtwork = null;
+		displayCapabilities = null;
+		inProgressUpdate = null;
+		queuedImageUpdate = null;
+		currentScreenData = null;
+		queuedUpdateListener = null;
+		pendingHMIListener = null;
+		inProgressListener = null;
+		hasQueuedUpdate = false;
+		isDirty = false;
+		pendingHMIFull = false;
+
+		// remove listeners
+		internalInterface.removeOnRPCNotificationListener(FunctionID.ON_HMI_STATUS, hmiListener);
+		internalInterface.removeOnSystemCapabilityListener(SystemCapabilityType.DISPLAY, onDisplayCapabilitiesListener);
+
+		// transition state
+		transitionToState(SHUTDOWN);
+	}
+
+	private void addListeners() {
+		// add listener
+		hmiListener = new OnRPCNotificationListener() {
+			@Override
+			public void onNotified(RPCNotification notification) {
+				currentHMILevel = ((OnHMIStatus)notification).getHmiLevel();
+				if (currentHMILevel == HMILevel.HMI_FULL){
+					if (pendingHMIFull){
+						Log.v(TAG, "Acquired HMI_FULL with pending update. Sending now");
+						pendingHMIFull = false;
+						sdl_update(pendingHMIListener);
+						pendingHMIListener = null;
+					}
+				}
+			}
+		};
+		internalInterface.addOnRPCNotificationListener(FunctionID.ON_HMI_STATUS, hmiListener);
+
+		// Add OnDisplayCapabilitiesListener to keep displayCapabilities updated
+		onDisplayCapabilitiesListener = new OnSystemCapabilityListener() {
+			@Override
+			public void onCapabilityRetrieved(Object capability) {
+				displayCapabilities = (DisplayCapabilities)capability;
+			}
+
+			@Override
+			public void onError(String info) {
+				Log.w(TAG, "DISPLAY Capability cannot be retrieved:");
+				displayCapabilities = null;
+			}
+		};
+		this.internalInterface.addOnSystemCapabilityListener(SystemCapabilityType.DISPLAY, onDisplayCapabilitiesListener);
+	}
+
+	// Upload / Send
+
+	protected void update(CompletionListener listener) {
+
+		// check if is batch update
+		if (batchingUpdates) {
+			return;
+		}
+
+		if (isDirty){
+			isDirty = false;
+			sdl_update(listener);
+		}
+	}
+	
+	private void sdl_update(CompletionListener listener){
+
+		// make sure hmi is not none
+		if (currentHMILevel == null || currentHMILevel == HMILevel.HMI_NONE){
+			Log.v(TAG, "Trying to send show on HMI_NONE, waiting for full");
+			pendingHMIFull = true;
+			if (listener != null){
+				pendingHMIListener = listener;
+			}
+			return;
+		}
+
+		Log.v(TAG, "Updating Text and Graphics");
+		if (inProgressUpdate != null){
```

Comment:
This variable seems to be updated from a few different methods. I would recommend some type of synchronization; adding the `volatile` keyword to the variable declaration should suffice.